src/c4/swap.t.hh

```diff
@@ -20,44 +17,54 @@
 #include "ds++/Assert.hh"
 
 namespace rtt_c4 {
-using std::vector;
 
 //---------------------------------------------------------------------------//
 // EXCHANGE
 //---------------------------------------------------------------------------//
 
 #ifdef C4_MPI
 
-template <class T>
-DLL_PUBLIC_c4 void determinate_swap(vector<unsigned> const &outgoing_pid,
-                                    vector<vector<T>> const &outgoing_data,
-                                    vector<unsigned> const &incoming_pid,
-                                    vector<vector<T>> &incoming_data, int tag) {
+template <typename T>
+DLL_PUBLIC_c4 void
+determinate_swap(std::vector<unsigned> const &outgoing_pid,
+                 std::vector<std::vector<T>> const &outgoing_data,
+                 std::vector<unsigned> const &incoming_pid,
+                 std::vector<std::vector<T>> &incoming_data, int tag) {
   Require(outgoing_pid.size() == outgoing_data.size());
   Require(incoming_pid.size() == incoming_data.size());
   Require(&outgoing_data != &incoming_data);
 
-  unsigned incoming_processor_count = incoming_pid.size();
-  unsigned outgoing_processor_count = outgoing_pid.size();
+  Check(incoming_pid.size() < UINT_MAX);
+  unsigned incoming_processor_count =
+      static_cast<unsigned>(incoming_pid.size());
+  Check(outgoing_pid.size() < UINT_MAX);
+  unsigned outgoing_processor_count =
+      static_cast<unsigned>(outgoing_pid.size());
 
   // This block is a no-op for with-c4=scalar.
   // Dito when the vectors are of zero-length.
   {
 
     // Post the asynchronous sends.
-    vector<C4_Req> outgoing_C4_Req(outgoing_processor_count);
+    std::vector<C4_Req> outgoing_C4_Req(outgoing_processor_count);
     for (unsigned p = 0; p < outgoing_processor_count; ++p) {
+      Check(outgoing_data[p].size() < INT_MAX);
```

Comment:
The signature for `rtt_c4::send_async` uses `int` for size and pid.  Since we are stuffing unsigned data (e.g.: `vector<T>.size()` or `unsigned int`) into a `signed int`, we must ensure that the unsigned data is less than the max allowed signed value.